### ISSUE TITLE
Is there a way to setup/teardown before running the load tests

### ISSUE THREAD

**Issue by sarvi** [locustio/locust#553]
This is usually stuff that needs to happen only once across before starting any clients/tasks
and once after all load test is complete.

I can use the quit event for the tear down i think.

Is there a start event that I can plugin into for setup ?

**Comment by cgoldberg** [comment]
you want to use `on_start` (noth that is run per virtual user)
see: http://docs.locust.io/en/latest/writing-a-locustfile.html#the-on-start-function
and the examples at: http://docs.locust.io/en/latest/quickstart.html

**Comment by sarvi** [comment]
I am looking for some way to invoke a setup() code before starting any virtual user.
Think of this as a way to setup/teardown the testbed/servers/etc. that all virtual user/client will be talking to. 

Right now, the only for me to do it, is to wrap locustio code with my own python wrapper, that does the setup/teardown, etc and then invoke locustio.main()

I don't think on_start() works that way does it ?

**Comment by cgoldberg** [comment]
> I don't think on_start() works that way does it ?

not really... it runs per user, so wouldn't really be useful for a global setup step.

**Comment by aldenpeterson-wf** [comment]
You should be able to add the code to the locustfile before you define the locust or anything else for your startup event.

Just make sure your code is setup to execute on import.

**Comment by MushuEE** [comment]
I need to do a singular setup to prime a database with information prior to the TaskSet execution. I need to access self.client to post resources... How would I access that variable prior to locust setup?

**Comment by payton** [comment]
Bumping this. I am in the same situation as @MushuEE 

I need to create resources before the load tests start and remove them after all load tests finish.

**Comment by cgoldberg** [comment]
see: https://docs.locust.io/en/stable/writing-a-locustfile.html#setups-teardowns-on-start-and-on-stop

**Comment by payton** [comment]
Right, but the setup/teardown functions under a locust are executed before/after an individual locust is hatched, correct? I was looking for a way to setup shared resources among all locusts and teardown those shared resources at the very end.

**Comment by MushuEE** [comment]
@payton you can overload the `setup()` method from `HttpLocust`
```
class WebsiteUser(HttpLocust):
    def setup(self):
        client = clients.HttpSession(base_url=self.host)
        client.post("/resources/", json=RESOURCE_1, headers=headers_with_auth)

    task_set = UserBehavior
    min_wait = 500
    max_wait = 900
```

I have not touched this for a little while now so I hope that helps

**Comment by payton** [comment]
@MushuEE where does the `events` object come from in the teardown function setup?

**Comment by GUSTAVOVARGASAVILA** [comment]
@sarvi / @payton, they could find how to solve their need?

**Comment by GUSTAVOVARGASAVILA** [comment]
It worked this way for me. Reading a file before starting:

class User(HttpUser):
    host = 'http://localhost:5000'        
    tasks = [UserBehavior]
    wait_time = between(1, 2)
    sock = None
    
    @events.test_start.add_listener
    def on_test_start(**kw):
        global USER_CREDENTIALS
        print("================================================================")
        if (USER_CREDENTIALS == None):
            with open('credentials.csv', mode='r') as file:
                reader = csv.reader(file)                
                USER_CREDENTIALS = [tuple(r) for r in reader]
                
    @events.test_stop.add_listener
    def on_test_start(**kw):
        print("test is stopping")

**Comment by TWoeste** [comment]
You can try to use the @event.init.add_listener
Here is an example [from the docs](https://docs.locust.io/en/stable/writing-a-locustfile.html#setups-teardowns-on-start-and-on-stop)

```from locust import events
from locust.runners import MasterRunner

@events.init.add_listener
def on_locust_init(environment, **kwargs):
    if isinstance(environment.runner, MasterRunner):
        print("I'm on master node")
    else:
        print("I'm on a worker or standalone node")```
